File: lab2/outputFiles/2.py

class Animal:
	_name = ''
	_sound = ''
	
	def __init__(self, animalName, animalSound):
		self._name = animalName
		self._sound = animalSound
	
	def makeSound(self):
		print(self._name, " makes sound: " ,self._sound)
	
	def eat(self):
		print(self._name, " is eating...\n", end='')
	
	def sleep(self):
		print(self._name, " is sleeping...\n", end='')
	
	def getName(self):
		return self._name
	def getSound(self):
		return self._sound
	
	def setName(self, newName):
		self._name = newName
	
	def displayInfo(self):
		print("Animal: " ,self._name, "\n", end='')

class Cat(Animal):
	__furColor = ''
	
	def __init__(self, catName, color = "gray"):
		self.__furColor = color
		self._name = catName
		self._sound = "Mewow!"
	
	def makeSound(self):
		print(self._name, " meows: " ,self._sound)
	
	def eat(self):
		print(self._name, " eats fish and drinks milk\n", end='')
	
	def climbTree(self):
		print(self._name, " climbs trees!\n", end='')
	
	def purr(self):
		print(self._name, " purrs: purrrrr...\n", end='')
	
	def displayInfo(self):
		print("Cat: " ,self._name, ", fur color: " ,self.__furColor, "\n", end='')
	
	def getFurColor(self):
		return self.__furColor

class Dog(Animal):
	__isTrained = False
	
	def __init__(self, dogName, trained = False):
		self.__isTrained = trained
		self._sound = "Woof!"
		self._name = dogName
	
	def makeSound(self):
		print(self._name, " barks: " ,self._sound)
	
	def eat(self):
		print(self._name, " eats bones and dry food\n", end='')
	
	def fetch(self):
		print(self._name, " fetches a stick!\n", end='')
	
	def train(self):
		if  not self.__isTrained:
			self.__isTrained = True
			print(self._name, " is now trained!\n", end='')
		else:
			print(self._name, " is already trained!\n", end='')
	
	def displayInfo(self):
		trained = {"yes" if self.__isTrained else "no"}
		print("Dog: " ,self._name, ", trained: " , trained , "\n", end='')
	
	def getIsTrained(self):
		return self.__isTrained

def animalShow(animal):
	animal.displayInfo()
	animal.makeSound()
	animal.eat()
	print("--------------------\n", end='')

if __name__ == '__main__':
	cat = Cat("Whiskers", "orange")
	dog = Dog("Buddy", False)
	
	print("=== Cats demonstration ===\n", end='')
	cat.displayInfo()
	cat.makeSound()
	cat.eat()
	cat.purr()
	cat.climbTree()
	print("\n", end='')
	
	print("=== Dogs demonstration ===\n", end='')
	dog.displayInfo()
	dog.makeSound()
	dog.eat()
	dog.fetch()
	dog.train()
	dog.train()
	print("\n", end='')
	
	print("=== Polymorphism demonstration ===\n", end='')
	animalShow(cat)
	animalShow(dog)
	

File: lab2/testFiles/2.cpp
#include <iostream>
#include <string>

class Animal {
protected:
    std::string name;
    std::string sound;

public:
    Animal(const std::string& animalName, const std::string& animalSound) {
        name = animalName;
        sound = animalSound;
    }

    virtual void makeSound() const {
        std::cout << name << " makes sound: " << sound << std::endl;
    }

    virtual void eat() const {
        std::cout << name << " is eating...\n";
    }

    virtual void sleep() const {
        std::cout << name << " is sleeping...\n";
    }

    std::string getName() const { 
        return name; 
    }
    std::string getSound() const { 
        return sound; 
    }

    void setName(const std::string& newName) { 
        name = newName; 
    }

    virtual void displayInfo() const {
        std::cout << "Animal: " << name << "\n";
    }
};

class Cat : public Animal {
private:
    std::string furColor;

public:
    Cat(const std::string& catName, const std::string& color = "gray"){
        furColor = color;
        name = catName;
        sound = "Mewow!";
    }

    void makeSound() const override {
        std::cout << name << " meows: " << sound << std::endl;
    }

    void eat() const override {
        std::cout << name << " eats fish and drinks milk\n";
    }

    void climbTree() const {
        std::cout << name << " climbs trees!\n";
    }

    void purr() const {
        std::cout << name << " purrs: purrrrr...\n";
    }

    void displayInfo() const override {
        std::cout << "Cat: " << name << ", fur color: " << furColor << "\n";
    }

    std::string getFurColor() const { 
        return furColor;
    }
};

class Dog : public Animal {
private:
    bool isTrained;

public:
    Dog(const std::string& dogName, bool trained = false){
        isTrained = trained;
        sound = "Woof!";
        name = dogName;
    }

    void makeSound() const override {
        std::cout << name << " barks: " << sound << std::endl;
    }

    void eat() const override {
        std::cout << name << " eats bones and dry food\n";
    }

    void fetch() const {
        std::cout << name << " fetches a stick!\n";
    }

    void train() {
        if (!isTrained) {
            isTrained = true;
            std::cout << name << " is now trained!\n";
        }
        else {
            std::cout << name << " is already trained!\n";
        }
    }

    void displayInfo() const override {
        std::string trained = (isTrained ? "yes" : "no");
        std::cout << "Dog: " << name << ", trained: " << trained << "\n";
    }

    bool getIsTrained() const { 
        return isTrained; 
    }
};

void animalShow(const Animal& animal) {
    animal.displayInfo();
    animal.makeSound();
    animal.eat();
    std::cout << "--------------------\n";
}

int main() {
    Cat cat("Whiskers", "orange");
    Dog dog("Buddy", false);

    std::cout << "=== Cats demonstration ===\n";
    cat.displayInfo();
    cat.makeSound();
    cat.eat();
    cat.purr();
    cat.climbTree();
    std::cout << "\n";

    std::cout << "=== Dogs demonstration ===\n";
    dog.displayInfo();
    dog.makeSound();
    dog.eat();
    dog.fetch();
    dog.train();
    dog.train();
    std::cout << "\n";

    std::cout << "=== Polymorphism demonstration ===\n";
    animalShow(cat);
    animalShow(dog);

    return 0;
}